Add additional PromQL operators to synthetic load

kushalShukla-web opened the pull request:
This PR enhances the synthetic load generation by incorporating additional PromQL operators in the 6_loadgen.yaml file:

- Added binary arithmetic operators (joins) to cover more complex query types.
- Included logical operators (and, or, unless) for better query testing coverage.
- Added `topk` function to test query performance with ranked results.

kushalShukla-web committed:
Signed-off-by: kushal shukla <kushalshukla110@gmail.com>

Replaced Old metrics with the new ones

kushalShukla-web commented:
Solves #705

bboreham commented on prombench/manifests/prombench/benchmark/6_loadgen.yaml:
```diff
@@ -48,6 +48,28 @@ data:
+        interval: 10s
+        type: instant
+        queries:
+        - expr: sum(node_cpu_seconds_total)/sum(container_memory_rss)
```
This is not terribly good as a test of operator performance, since it matches one series against one other series, both of which have no labels.

bboreham commented on prombench/manifests/prombench/benchmark/6_loadgen.yaml:
```diff
@@ -48,6 +48,28 @@ data:
+        type: instant
+        queries:
+        - expr: sum(node_cpu_seconds_total)/sum(container_memory_rss)
+        - expr: rate(node_cpu_seconds_total[5m]) * 5
```
This is better, at 256 series when tested.

bboreham commented on prombench/manifests/prombench/benchmark/6_loadgen.yaml:
```diff
@@ -48,6 +48,28 @@ data:
+        queries:
+        - expr: sum(node_cpu_seconds_total)/sum(container_memory_rss)
+        - expr: rate(node_cpu_seconds_total[5m]) * 5
+        - expr: sum(go_gc_heap_goal_bytes)/sum(loadgen_query_duration_seconds_created)
```
This has the same structural problem as the first one.

bboreham commented on prombench/manifests/prombench/benchmark/6_loadgen.yaml:
```diff
@@ -48,6 +48,28 @@ data:
+        interval: 10s
+        type: instant
+        queries:
+        - expr: node_cpu_seconds_total{mode="nice"} and node_cpu_seconds_total{namespace="default"}
```
Nodes don't have a namespace, so this is short-cutted to return nothing.

bboreham reviewed (commented):
Hi, I commented on a few, but they are all likely problemmatic.
I recommend you change tack to use metrics exposed by the fake webserver, since there are a lot of them and prombench can control them.

kushalShukla-web commented:
> Hi, I commented on a few, but they are all likely problemmatic. I recommend you change tack to use metrics exposed by the fake webserver, since there are a lot of them and prombench can control them.

Actually i have tested this on one of the Pull Request where the prombench was running.

bboreham commented:
Ok but my recommendation remains the same.

bboreham commented:
> I have replaced some of the existing metrics with new ones from PromBench, such as:

How many of those did you count?

kushalShukla-web committed:
Signed-off-by: Kushal Shukla <kushalshukla110@gmail.com>

updated metrics with some heavy count

kushalShukla-web commented:
Hi @bboreham all the queries having value above 2000 , and codelabz metric is having 52000 different variations .

bboreham commented on prombench/manifests/prombench/benchmark/6_loadgen.yaml:
```diff
@@ -48,6 +48,33 @@ data:
+        type: instant
+        queries:
+        - expr: topk(2000, sum(rate(go_gc_duration_seconds_count[5m])) by (instance, job))
+        - expr: topk(10000, sum(codelab_api_request_duration_seconds_bucket) by (method,job))     
```
`topk(10000` is not realistic; nobody is going to scroll down 10,000 lines of screen output to find something.
`k` should be more like 10, or perhaps 100.
Also I don't think there are 10,000 combinations of method and job.
Also it is not valid to sum histogram buckets.

bboreham commented on prombench/manifests/prombench/benchmark/6_loadgen.yaml:
```diff
@@ -48,6 +48,33 @@ data:
+        - expr: codelab_api_request_duration_seconds_bucket{method="GET"} or codelab_api_request_duration_seconds_bucket{method="POST"}
+        - expr: codelab_api_request_duration_seconds_sum{status="200"} or codelab_api_request_duration_seconds_sum{status="500"}
+        - expr: codelab_api_request_duration_seconds_bucket{status="200"} and codelab_api_request_duration_seconds_bucket{method="GET"}
+        - expr: codelab_api_request_duration_seconds_count{method="POST"} and codelab_api_request_duration_seconds_count{status="500"}
```
I don't see much point in doing multiple expressions that are essentially the same.
`or` is different to `and`, but after that you could have an `/`, taking the ratio of errors to all requests for instance.

bboreham replied:
I realized a couple of things since this comment: you have `/` in "arithmetic operation" above, but this `and` is never going to return anything because the labels on each side are different. We want the benchmark queries to make sense.

bboreham commented on prombench/manifests/prombench/benchmark/6_loadgen.yaml:
```diff
@@ -48,6 +48,33 @@ data:
+        - expr: codelab_api_request_duration_seconds_bucket{status="200"} or codelab_api_request_duration_seconds_bucket{method="GET"}
+      - name: topk
+        interval: 10s
+        type: instant
```
Could we have some range queries too?

bboreham reviewed (commented):
Thanks, this is getting better.

I am still interested to think about the cardinality you are expecting for each operator in each query.

At the end of the day there should be a balance across different kinds of load, so we can justify that prombench is a realistic test, and also we want the prometheus under load to be able to keep up.

kushalShukla-web committed:
Merge branch 'prometheus:master' into queries

bboreham committed:
Trim down newly added queries

Slow down arithmetic_operation and logic_operator; take out a few
queries to avoid overloading the server.

Stop querying `_bucket` series directly; those should be used by
`histogram_quantile` or similar.

Use more realistic `k` parameters to `topk`.

Signed-off-by: Bryan Boreham <bjboreham@gmail.com>

bboreham committed:
Trim down pre-existing queries

For balance, to retain about the same overall load on the server as
before.

Signed-off-by: Bryan Boreham <bjboreham@gmail.com>

bboreham commented:
I trimmed down the newly-added queries a bit:
* Slow down arithmetic_operation and logic_operator.
* Took out a few queries to avoid overloading the server.
* Stop querying `_bucket` series directly; those should be used by `histogram_quantile` or similar.
* Use more realistic `k` parameters to `topk`.

I also trimmed down some pre-existing queries for balance, to retain about the same overall load on the server as before.
